Add setting to control adding "(guest)" indicator to user names.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit adds the new setting.
- Second commit is for adding indicator based on the setting in various places like pills, typeaheads, message feed, etc.
- Third commit is for adding indicator to mention pill.
- Fourth commit is to add indicator to sender names in inbox view.
- Fifth commit is to add indicator to narrow titles.
- Sixth commit is to update `get_display_full_name` function to add indicator at various places.
- Seventh commit is to add indicator to compose box placeholder.
- Last commit is to add indicator to sender avatar tooltips in recent topics view.

Fixes: <!-- Issue link, or clear description.--> #26700

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
![Peek 2023-09-18 16-18](https://github.com/zulip/zulip/assets/35494118/62e6c297-672e-425a-824f-5648042d7222)

sahil839 commented on web/src/compose_ui.js:
```diff
@@ -244,6 +244,9 @@ export function compute_placeholder_text(opts) {
             .map((recipient) => {
                 const user = people.get_by_email(recipient);
+                if (people.should_add_guest_user_indicator(user.user_id)) {
+                    return $t({defaultMessage: "{name} (guest)"}, {name: user.full_name});
```
I guess this will happen in a couple of other cases as well due to the commit modifying `get_display_full_names`.

timabbott replied:
I think this is probably OK in that the only types of things I expect a translator to do with this string is translate the single word "guest", replace parens with a language-specific equivalent, and maybe swap the order for RTL languages. I think it's OK for now; this is limited to a setting that isn't going to be enabled in every organization.

timabbott commented on web/templates/user_full_name.hbs:
```diff
@@ -0,0 +1,5 @@
+{{#tr}}{name} (guest){{/tr}}
+{{else}}
+{{name}}
+{{/if}}
```
Yeah, I like that -- it'll also make it way easier to style the `(guest)` part differently from text if we decide to do that.

timabbott commented on web/src/filter.js:
```diff
@@ -706,9 +706,14 @@ export class Filter {
+                if (people.should_add_guest_user_indicator(person.user_id)) {
+                    return $t({defaultMessage: "{name} (guest)"}, {name: person.full_name});
+                }
                 return people.get_by_email(email).full_name;
```
I think this should be `return person.full_name`

timabbott commented on web/src/filter.js:
```diff
@@ -727,6 +732,11 @@ export class Filter {
+            }
+
             return $t(
                 {defaultMessage: "Messages sent by {sender}"},
```
I feel like this pattern feels very awkward/ugly -- ideally, we'd not create duplicates of strings like this for the guest case, but instead just have the `{sender}` part be substituted with a translation of `{name} (guest}`, like what you generate in https://github.com/zulip/zulip/pull/26766/files#r1328971956. Care may need to be taken to avoid double-escaping.

It's likely worth a thread in #translating about what to do here. But I'm thinking that for use cases that for whatever reason don't support styling the "guest" part differently via say including the template, or something fancier, we probably want to avoid needing a separate duplicate of a string just for this feature if at all possible.

sahil839 commented on web/src/message_list_view.js:
```diff
@@ -223,7 +238,8 @@ function populate_group_from_message_container(group, message_container) {
         group.all_visibility_policies = user_topics.all_visibility_policies;
     } else if (group.is_private) {
         group.pm_with_url = message_container.pm_with_url;
-        group.display_reply_to = message_store.get_pm_full_names(message_container.msg);
```
I added new function instead of using `get_pm_full_names` because the string returned by `get_pm_full_names` and `get_display_full_name`, which is used by `get_pm_full_names` to get the user names from IDs, is used as text and not html in many places. So changing the code in `get_display_full_name` to have '<i>` tag around "guest" results in the text being shown as `Polonius <i>(guest)</i>`.

We can probably fix those while being careful about escaping and translation but would be better to do that in a separate PR.

karlstolley replied:
Good catch, @sahil839. One thing to note is that pills and sidebars that use flexbox are often having to work with usernames as plain text nodes. So an even better solution might be to have an HTML element surrounding both (e.g., `<span class="some-new-unique-user-class">Polonius <i>(guest)</i></span>`). That would likely work better with any existing flexboxes (which create an anonymous box around any text nodes), and prevent the `<i>` portion from becoming its own flex item, which could make for a whole lot of weird bugs on flexboxes alone.

sahil839 replied:
@karlstolley  So, if I am understanding correctly, this means that we should make sure in cases where flexbox layout is used the name and "<i>(guest)</i>" element should be inside a single HTML element if it is direct children of a flex container.

karlstolley replied:
Yes. And you'd have to set it up so that the `text-overflow` properties were on that `.some-new-unique-user-class`, because it would be the username text-node that would get the ellipsis, while keeping the `<i>(guest)</i>` portion always visible. It might take some tinkering to get that to work as expected inside existing flex containers, but I'm always happy to jam on this kind of problem if you think I could be of some help.

sahil839 commented:
Added commit to only cut-off names in case of long names and show the indicator in right sidebar and message feed.

As a result done some changes for `user_full_name` component which is used in user profile popover and modal as well but currently the indicator is also cut off in these places for long names since that would need some additional CSS changes and we anyways show the role in those places. Can do those changes as a follow-up if we need to make sure that indicator is shown for long names as well.

@karlstolley can you have a look at the latest commit? I have done some changes for me-messages as well, but am not sure whether there exists a better solution.

sahil839 commented:
Updated the PR, but could not fix the "/me" message case. The fix that I thought previously (by setting `display: flex` along with `flex-wrap: wrap` for `.message_sender` [here](https://github.com/zulip/zulip/blob/35e4be0b683373e04d4343a4cf34361082bc1ef6/web/styles/message_row.css#L343)) had a bug where either the complete message was on single line or the sender name on one line and the message on second even if there is some space after the sender name to have a part of the message. I am sharing the GIF for the current behavior below. @karlstolley any suggestions on how to fix the "/me" message case.
![guest-indicator-long-names](https://github.com/zulip/zulip/assets/35494118/029ef3a0-63e7-4552-a4ee-50a90f149f5b)

karlstolley commented:
I must be doing something wrong--I've pulled down your PR and reprovisioned, but I don't seem to see the `(guest)` indicator anywhere in the browser.

karlstolley commented:
Okay. This is a tricky problem, but here's the fix:
<img width="598" alt="Screenshot 2023-10-02 at 2 19 29 PM" src="https://github.com/zulip/zulip/assets/170719/1b2a41ed-c904-4406-87cc-38179f5a77ab">

Inside the .me-message, that `inline_status_emoji` class should be set to display as an inline flex. We hold its width to 100%, so that setting the `white-space: nowrap` on the username doesn't hold the entire flex container out beyond the limits of the area allotted in the message grid:

![Screen Shot 2023-10-02 at 14 19 54](https://github.com/zulip/zulip/assets/170719/f3972709-8a5b-49e8-9c3a-e3fb8488d7bb)

Note that it looks like all of your flex containers like this need a `column-gap` set; the whitespace you're trying to set gets collapsed by flexbox.

sahil839 commented:
Thanks for the suggestion @karlstolley. I was trying this as well, but was missing setting width. Updated the PR with the fix suggested.

Also updated the PR to fix the space before the indicator. Note that I still need to work on live-update part.

alya commented:
I think it's looking pretty good! Can we italicize "(guest)" in the left sidebar, if it's not too hard?

![Screenshot 2023-10-03 at 2 04 20 PM](https://github.com/zulip/zulip/assets/2090066/549b0ca8-9341-429a-89b2-4ed7950de30a)

alya commented:
I guess it's also not italicized in Inbox and Recent conversations. Probably would be best to italicized there as well, but I don't think it's critical.

timabbott commented:
Do we want this to live in the `User identity`  section of "organization permissions" rather than at the bottom of "organization settings"?

![image](https://github.com/zulip/zulip/assets/2746074/33e2eaf9-3ebe-4c5e-95f6-8cf1d81bf030)

timabbott commented on web/src/message_list_view.js:
```diff
@@ -183,6 +183,21 @@ function set_topic_edit_properties(group, message) {
+            return _.escape(person.full_name);
+        })
+        .join(", ");
+}
```
I really don't like using `{{{` as every time we use it adds security risk if it ends up being used incorrectly. Is there a reason we can't just pass objects into the recipient row rendering and have the loop be in Handlebars?

andersk replied:
Are you asking how to do `{{#unless @last}},{{/unless}}` (documentation: [`@last`](https://handlebarsjs.com/api-reference/data-variables.html#last))?

sahil839 replied:
No. I am basically not sure about how to handle translation here. I guess we cannot just tag "You and" for translation and then add comma separated names. And I think we also cannot add handlebars inside `{[#tr}}`.

sahil839 commented:
Updated the PR to resolve merge conflicts.

Regarding @alya's feedback above - we can italicize the indicator in all the mentioned places, but the function used for displaying those names is used in many places. We can add a new function to use in these places, but that would lead to duplication, so would be better to just refactor them all if we want to avoid duplication. I am working on it and it could take a couple of days. We can merge this PR if we are fine to do those changes as follow-ups though or else I will try to complete them soon.

timabbott commented:
I would be very happy to see any additional work on italicizing done as follow-ups; those refactoring changes are going to be way easier to review individually if complicated. So I'd prefer we work to resolve https://github.com/zulip/zulip/pull/26766#pullrequestreview-1656333808 and get this merged, and then we can do follow-up work on additional italicization polish.

alya commented:
That plan sounds good to me! Let's get the initial feature merged, and file an issue for any additional places we are aware of where we'd like to add italics. Those follow-ups will be a lower priority than some of the other projects you have on your place.

sahil839 commented on web/templates/recipient_row.hbs:
```diff
@@ -94,9 +94,17 @@
+            {{#tr}}You and <z-user-names></z-user-names>
+            {{#*inline "z-user-names"}}
+                {{#each recipient_users}}
+                    {{full_name}}{{#if should_add_guest_user_indicator}}&nbsp;<i>({{t 'guest'}})</i>{{/if}}{{#unless @last}},{{/unless}}
```
Have done this changes, to avoid using `{{{` but now we have translated guest text inside `#tr` tag.

timabbott replied:
Great, thanks!

alya commented:
> Do we want this to live in the User identity section of "organization permissions" rather than at the bottom of "organization settings"?

I like that idea. It's not quite a permission setting, but "User identity" is very appropriate, and the bottom of org settings is getting rather long. Maybe let's put it at the end of that "User identity" subsection?

alya commented:
> Also, added a commit to show indicator for muted users as well, but was not sure about whether we want to do this, so did this change in a separate commit.

Could you please post a screenshot?

sahil839 commented:
Updated the PR to move the setting to "User Identity" section. And I have not added live-udpate code for now in this PR, since this PR is already very long. Can add as a follow-up.

alya commented:
Thanks! Displaying "(guest)" for muted users seems reasonable to me. It's not controlled by the muted user, so I don't really see an opportunity for harassment by showing it.

timabbott commented on web/src/message_list_view.js:
```diff
@@ -183,6 +183,28 @@ function set_topic_edit_properties(group, message) {
+    });
+
+    function compare_by_name(a, b) {
+        return a.full_name < b.full_name ? -1 : a.full_name > b.full_name ? 1 : 0;
```
Please start a #design thread on this question.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14080,6 +14086,14 @@ paths:
+
+                          Whether to add "(guest)" indicator to user names in UI.
+
+                          **Changes**: New in Zulip 8.0 (feature level 216).
```
This probably deserves a bit of extra exposition; I might just add it as part of merging.

timabbott commented on web/src/message_list_view.js:
```diff
@@ -183,6 +183,28 @@ function set_topic_edit_properties(group, message) {
     }
 }
 
+function get_users_for_recipient_row(message) {
```
We should probably move this function to `people.js` to sit alongside `get_display_full_name` -- it's best to avoid having multiple similar algorithms not next to each other in the codebase.

timabbott commented:
I made the following changes to the first commit to improve the label for the setting as well as its API documentation. Regarding the API documentation, and important principle for any of these settings is to clearly precisely what software is supposed to do the work -- so in this case, we specify clearly that this is a thing clients are supposed to do.

``` diff
diff --git a/web/src/admin.js b/web/src/admin.js
index 9e61f4af21..4e1110b9d9 100644
--- a/web/src/admin.js
+++ b/web/src/admin.js
@@ -62,7 +62,7 @@ const admin_settings_label = {
         defaultMessage: "Users can always disable their personal read receipts.",
     }),
     realm_enable_guest_user_indicator: $t({
-        defaultMessage: 'Add "(guest)" to names of guest users',
+        defaultMessage: "Display “(guest)” after names of guest users",
     }),
 };
 
diff --git a/zerver/models.py b/zerver/models.py
index 5d5698cc97..461405bd1a 100644
--- a/zerver/models.py
+++ b/zerver/models.py
@@ -712,8 +712,8 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
     # they will not be available regardless of users' personal settings.
     enable_read_receipts = models.BooleanField(default=False)
 
-    # Whether to add "(guest)" indicator to user names.
-    enable_guest_user_indicator = models.BooleanField(default=False)
+    # Whether clients should display "(guest)" after names of guest users.
+    enable_guest_user_indicator = models.BooleanField(default=True)
 
     # Duplicates of names for system group; TODO: Clean this up.
     ADMINISTRATORS_GROUP_NAME = "role:administrators"
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 4a8881aefe..02ded97178 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -4428,7 +4428,8 @@ paths:
                                     enable_guest_user_indicator:
                                       type: boolean
                                       description: |
-                                        Whether to add "(guest)" indicator to user names in UI.
+                                        Whether clients should display "(guest)" after the names of
+                                        guest users to prominently highlight their status.
 
                                         **Changes**: New in Zulip 8.0 (feature level 216).
                                   additionalProperties: false
@@ -14091,7 +14092,8 @@ paths:
                         description: |
                           Present if `realm` is present in `fetch_event_types`.
 
-                          Whether to add "(guest)" indicator to user names in UI.
+                          Whether clients should display "(guest)" after the names of
+                          guest users to prominently highlight their status.
 
                           **Changes**: New in Zulip 8.0 (feature level 216).
                       zulip_plan_is_not_limited:
```

timabbott commented:
Also, changes the default to `True`; my sense as to how this feature will be used is that probably most organizations will want it, though there will probably be some that don't; it's also more discoverable default-enabled.

With those changes, marked this to merge once CI passes. @alya can you take of arranging corresponding /help/ changes.
